File: frontend/features/exercise-import/services/exerciseService.ts
import api from '@/lib/services/axios-instance';
import { ApiResponse, PageResponse } from '@/lib/types';
import {
    CreateExerciseRequest,
    Exercise,
    ExerciseListItemResponse,
    ExerciseStatus,
    ImportExerciseRequest,
    ImportPreviewResponse,
    AssignExerciseRequest,
    TutorStudentSummaryResponse,
} from '../types/exercise.types';

export const exerciseService = {
    /**
     * Parse text content to preview exercise import
     */
    importPreview: async (content: string, classId?: string): Promise<ImportPreviewResponse> => {
        const request: ImportExerciseRequest = { content, classId };
        const response = await api.post<ApiResponse<ImportPreviewResponse>>('/exercises/import/text', request);
        return response.data.data;
    },

    /**
     * Create a new exercise
     */
    create: async (data: CreateExerciseRequest): Promise<Exercise> => {
        const response = await api.post<ApiResponse<Exercise>>('/exercises', data);
        return response.data.data;
    },

    /**
     * Update an existing exercise
     */
    update: async (id: string, data: CreateExerciseRequest): Promise<Exercise> => {
        const response = await api.put<ApiResponse<Exercise>>(`/exercises/${id}`, data);
        return response.data.data;
    },

    /**
     * Get exercise by ID
     */
    getById: async (id: string): Promise<Exercise> => {
        const response = await api.get<ApiResponse<Exercise>>(`/exercises/${id}`);
        return response.data.data;
    },

    /**
     * List exercises with optional filters and pagination
     */
    getAll: async (classId?: string, status?: ExerciseStatus, search?: string, page = 0, size = 10): Promise<PageResponse<ExerciseListItemResponse>> => {
        const params: any = { page, size };
        if (classId) params.classId = classId;
        if (status) params.status = status;
        if (search) params.search = search;

        const response = await api.get<ApiResponse<PageResponse<ExerciseListItemResponse>>>('/exercises', { params });
        return response.data.data;
    },

    /**
     * Delete an exercise
     */
    delete: async (id: string): Promise<void> => {
        const response = await api.delete<ApiResponse<void>>(`/exercises/${id}`);
        return response.data.data;
    },

    /**
     * Assign exercise to a specific student
     */
    assign: async (exerciseId: string, data: AssignExerciseRequest): Promise<void> => {
        const response = await api.post<ApiResponse<void>>(`/exercises/${exerciseId}/assign`, data);
        return response.data.data;
    },

    /**
     * Get exercises assigned to the current student
     */
    getAssigned: async (page = 0, size = 10): Promise<PageResponse<ExerciseListItemResponse>> => {
        const response = await api.get<ApiResponse<PageResponse<ExerciseListItemResponse>>>('/exercises/assigned', { params: { page, size } });
        return response.data.data;
    },

    /**
     * Get student summary list for tutor dashboard
     */
    getTutorStudentSummaries: async (page = 0, size = 10): Promise<PageResponse<TutorStudentSummaryResponse>> => {
        const response = await api.get<ApiResponse<PageResponse<TutorStudentSummaryResponse>>>('/tutor/dashboard/students', { params: { page, size } });
        return response.data.data;
    },

    /**
     * Get exercises assigned to a specific student (For Tutors/Admins)
     */
    getAssignedByStudentId: async (studentId: string, page = 0, size = 10): Promise<PageResponse<ExerciseListItemResponse>> => {
        const response = await api.get<ApiResponse<PageResponse<ExerciseListItemResponse>>>(`/exercises/assigned/${studentId}`, { params: { page, size } });
        return response.data.data;
    },
};

File: backend/src/main/java/com/tutor_management/backend/modules/exercise/service/ExerciseService.java
package com.tutor_management.backend.modules.exercise.service;

import com.tutor_management.backend.modules.exercise.domain.ExerciseStatus;
import com.tutor_management.backend.modules.exercise.dto.request.CreateExerciseRequest;
import com.tutor_management.backend.modules.exercise.dto.request.ImportExerciseRequest;
import com.tutor_management.backend.modules.exercise.dto.response.ExerciseListItemResponse;
import com.tutor_management.backend.modules.exercise.dto.response.ExerciseResponse;
import com.tutor_management.backend.modules.exercise.dto.response.ImportPreviewResponse;
import com.tutor_management.backend.modules.exercise.dto.response.TutorStudentSummaryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.List;

/**
 * Primary business orchestration layer for the Exercise module.
 * Manages resource lifecycles, student assignments, and integration with the parsing engine.
 */
public interface ExerciseService {
    
    /**
     * Provides a non-persistent preview of how a raw text block will be structured.
     * 
     * @param request Contains the raw content and optional target class mapping.
     * @return Structured preview data for tutor verification.
     */
    ImportPreviewResponse previewImport(ImportExerciseRequest request);
    
    /**
     * Persists a new exercise resource in the system.
     * 
     * @param request Full exercise definition (metadata + questions).
     * @param teacherId UUID of the authoring staff member.
     * @return The saved and ID-assigned exercise resource.
     */
    ExerciseResponse createExercise(CreateExerciseRequest request, String teacherId);
    
    /**
     * Updates an existing exercise. Replaces the sub-resource graph (questions/options).
     * 
     * @param id Target exercise UUID.
     * @param request Updated definition.
     * @param teacherId UUID of the performing staff member (for ownership check).
     * @return The updated exercise resource.
     */
    ExerciseResponse updateExercise(String id, CreateExerciseRequest request, String teacherId);
    
    /**
     * Fetches details for a specific exercise, including full question content.
     */
    ExerciseResponse getExercise(String id);
    
    /**
     * Retrieves a filtered list of exercises for the management library.
     * 
     * @param classId Optional class filter.
     * @param status Optional draft/published status filter.
     * @param pageable Pagination and sorting metadata.
     * @return Summary projections of matching resources.
     */
    Page<ExerciseListItemResponse> listExercises(String classId, ExerciseStatus status, String search, Pageable pageable);
    
    /**
     * Permanently removes an exercise and all its associated data (submissions, assignments).
     * 
     * @param id Target exercise UUID.
     * @param teacherId UUID of the performing staff member (for ownership check).
     */
    void deleteExercise(String id, String teacherId);

    /**
     * Links a published exercise to a specific student for completion.
     * 
     * @param exerciseId Source resource UUID.
     * @param studentId Target student recipient.
     * @param tutorId Authorizing staff member.
     * @param deadline Optional specific cutoff time.
     */
    void assignToStudent(String exerciseId, String studentId, String tutorId, LocalDateTime deadline);

    /**
     * Retrieves assigned materials for a student with pagination support.
     */
    Page<ExerciseListItemResponse> listAssignedExercises(String studentId, Pageable pageable);

    /**
     * Aggregates completion statistics for students with pagination support.
     */
    Page<TutorStudentSummaryResponse> getStudentSummaries(Pageable pageable);
}
